packages/babel-plugin-transform-react-jsx/src/transform-experimental.js

```diff
@@ -0,0 +1,753 @@
+import esutils from "esutils";
+import jsx from "@babel/plugin-syntax-jsx";
+import {
+  isModule,
+  addNamespace,
+  addNamed,
+  addDefault,
+} from "@babel/helper-module-imports";
+
+// These are all the valid auto import types (under the config autoImport)
+// that a user can specify
+const IMPORT_TYPES = {
+  none: "none", // default option. Will not import anything
+  require: "require", // var _react = require("react");
+  namespace: "namespace", // import * as _react from "react";
+  defaultExport: "defaultExport", // import _default from "react";
+  namedExports: "namedExports", // import { jsx } from "react";
+};
+
+const JSX_AUTO_IMPORT_ANNOTATION_REGEX = /\*?\s*@jsxAutoImport\s+([^\s]+)/;
+const JSX_IMPORT_SOURCE_ANNOTATION_REGEX = /\*?\s*@jsxImportSource\s+([^\s]+)/;
+
+// We want to use React.createElement, even in the case of
+// jsx, for <div {...props} key={key} /> to distinguish it
+// from <div key={key} {...props} />. This is an intermediary
```

Comment:
Can you explain why this is necessary?

When I try to reason about it, it seems like the current implementation is the opposite of what should happen. We can reliably tell that `<div {...props} key={KEY} />` is supposed to be `KEY`. In `<div key={KEY} {...props} />`, we can't tell if the key is `KEY` or was passed in from the spread props. So to maintain the current behavior while you deprecate, you should be using `createElement` only in the second case.

Reply:
For `<div key={key} {...props} />` we'll look at `props` to determine if it has a key and if so override the key passed in. So we preserve the behavior by keeping the runtime behavior + warn and then later change the runtime behavior.

For `<div {...props} key={key} />` we don't know if it is `<div {...props} key={undefined} />` and we can't distinguish that from `<div {...props} />` at runtime because they both pass `undefined` to the `key` argument.

Reply:
It makes sense :+1:
The comment should probably be updated to mention that it's needed for the `undefined` case.

<details>
<summary>Half joking</summary>

> because they both pass undefined to the key argument.

`arguments.length`? :stuck_out_tongue: 
</details>